src/common-ssh/ssh.c

```diff
@@ -412,6 +428,84 @@ static int guac_common_ssh_authenticate(guac_common_ssh_session* common_session)
 
 }
 
+/**
+ * Verifies if given algorithms are supported by libssh2 and
+ * makes the |prefs| string to use with libssh2_session_method_pref().
+ * 
+ * If the verification fails, all preferred algorithms |algs| are
+ * added to the |prefs| string.
+ *
+ * @param client
+ *     The Guacamole client that is using SSH.
+ * 
+ * @param session
+ *     The session associated with the user to be authenticated.
+ *
+ * @param method_type
+ *      One of the libssh2 Method Type constants for libssh2_session_method_pref().
+ * 
+ * @param algs
+ *      A preferred list of algorithms, for example FIPS_COMPLIANT_CIPHERS.
+ * 
+ * @return
+ *     The |prefs| string to use with libssh2_session_method_pref().
+ */
+static char *get_method_pref_string(guac_client* client, LIBSSH2_SESSION* session,
+        int method_type, const char** algs) {
+
+    /* Initial size of the string. It will be reallocated later if not enough. */
+    int algs_str_size = 1024;
+    char* algs_str = guac_mem_alloc(algs_str_size);
+    
+    algs_str[0] = 0;
+
+    /* Request the list of supported algorithms/cyphers from libssh2. */
+    const char **supported_algs;
+    int supported_algs_count =
+        libssh2_session_supported_algs(session, method_type, &supported_algs);
+
+    /* This is not a show stopper. */
+    if (supported_algs_count <= 0) {
+        guac_client_log(client, GUAC_LOG_WARNING,
+            "Could not verify if preferred algorithms/ciphers are supported by libssh2");
+    }
+
+    int i = 0;
+    while (algs[i]) {
+        bool found = false;
+        /* Check if the algorithm is found in the libssh2 supported list. */
+        for (int j = 0; j < supported_algs_count; j++) {
+             if (strcmp(algs[i], supported_algs[j]) == 0) {
+                found = true;
+                break;
+            }
+        }
```

Comment:
I'm wary of doing a brute-force search in a loop like this, but if we're certain the list cannot grow to any substantial length, then this is fine.

If the list might grow to become large, we should consider sorting the list and doing a `bsearch()`.